README.md
# Vedran daemon

> Daemon application for interacting with vedran load balancer.

### Get `vedran-daemon` package
1. Install [Golang](https://golang.org/doc/install) **1.13 or greater**
2. Run the command below
```
go get github.com/NodeFactoryIo/vedran-daemon
```
3. Run vedran-daemon from your Go bin directory. For linux systems it will likely be:
```
~/go/bin/vedran-daemon
```
Note that if you need to do this, you probably want to add your Go bin directory to your $PATH to make things easier!

## Usage

```
$ ./vedran-daemon -h
Register vedran-daemon with load balancer and start sending telemetry

Usage:
  vedran-daemon [flags]

Flags:
  -h, --help                    help for vedran-daemon
      --id string               Vedran-daemon id string (required)
      --lb string               Target load balancer url (required)
      --log-file string         Path to logfile. If not set defaults to stdout
      --log-level string        Level of logging (eg. debug, info, warn, error) (default "info")
      --node-metrics string     Polkadot node metrics url (default "localhost:9615")
      --node-rpc string         Polkadot node rpc url (default "localhost:9933")
      --node-ws string         Polkadot node ws url (default "localhost:9944")
      --payout-address string   Payout address to which reward tokens will be sent (required)
```
## Starting daemon

First download latest prebuilt binary of vedran daemon [from releases](https://github.com/NodeFactoryIo/vedran-daemon/releases) and binary for [node](https://github.com/paritytech/polkadot/releases).
Daemon is expected to be started in conjuction with node and will wait for node if it is unavailable.


### Node
For starting node see [instructions](https://github.com/paritytech/polkadot/blob/master/README.md)

**NOTE node should be started with rpc cors disabled

Example:
`
  ./polkadot --rpc-cors=all
`

### Daemon
Daemon is started by invoking binary.

For example:
```
  ./vedran-daemon-linux-amd64 --id UuPrCMnkni --lb https://load-balancer.com --payout-address 15MCkjt3B59dNo5reMCWWpxY8QB8VpEbYLo2xHEjuuWsSmTU
```

It will register to load balancer and start sending pings and metrics to load balancer.
Port forwarding to local node is not needed and node and daemon can be in a private network because [http tunnel](https://en.wikipedia.org/wiki/HTTP_tunnel) is created between
node and load balancer on registration which communicate via daemon used as a proxy server.

### Required flags

`--id` - id string by which load balancer will distinguish between nodes - **CAUTION** this should be a unique string and should not be shared

`--lb` - public url of vedran load balancer

`--payout-address` - address of wallet to which reward tokens should be set - **CAUTION** - use valid address depending on network

### Other flags

`--log-level` - log level (debug, info, warn, error) - **DEFAULT** [error]

`--log-file` - path to file in which logs will be saved - **DEFAULT** [stdout]

`--node-metrics` - local url to node metrics - **DEFAULT** [http://localhost:9615]

`--node-rpc` - local url to node rpc endpoint - **DEFAULT** [http://localhost:9933]

`--node-ws` - local url to node websocket rpc endpoint - **DEFAULT** [http://localhost:9944]

## Development


### Clone

```bash
git clone git@github.com:NodeFactoryIo/vedran-daemon.git
```

### Lint
[Golangci-lint](https://golangci-lint.run/usage/install/#local-installation) is expected to be installed.

```bash
make lint
```

### Build

```bash
make build
```

### Test

```bash
make test
```

Run daemon app with `go run main.go [command]`.

More about different _commands_ can be found in [Usage](#Usage).

## License

This project is licensed under Apache 2.0:
- Apache License, Version 2.0, [LICENSE](LICENSE) or http://www.apache.org/licenses/LICENSE-2.0


CHANGELOG.md
# Changelog

## [v0.3.4](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.3.4)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.3.3...v0.3.4)

### Fixed
- Fix socket and request memory leaks [\#63](https://github.com/NodeFactoryIo/vedran-daemon/pull/63) ([mpetrun5](https://github.com/mpetrun5))

### Added

### Changed

## [v0.3.3](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.3.3)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.3.2...v0.3.3)

### Fixed

### Added
- Add ws endpoint check on start [\#62](https://github.com/NodeFactoryIo/vedran-daemon/pull/62) ([MakMuftic](https://github.com/MakMuftic))

### Changed

## [v0.3.2](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.3.2)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.3.1...v0.3.2)

### Fixed
- Add target block to metrics [\#58](https://github.com/NodeFactoryIo/vedran-daemon/pull/58) ([MakMuftic](https://github.com/MakMuftic))

### Added

### Changed

## [v0.3.1](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.3.1)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.3.0...v0.3.1)

### Fixed
- Fix displaying configuration [\#53](https://github.com/NodeFactoryIo/vedran-daemon/pull/53) ([MakMuftic](https://github.com/MakMuftic))

### Added

### Changed

## [v0.3.0](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.3.0)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.2.0...v0.3.0)

### Fixed

### Added
- Added websocket support [\#49](https://github.com/NodeFactoryIo/vedran-daemon/pull/49) ([mpetrun5](https://github.com/mpetrun5))

### Changed


## [v0.2.0](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.2.0)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.1.2...v0.2.0)

### Fixed

### Added
- Added banner on starting vedran-daemon [\#32](https://github.com/NodeFactoryIo/vedran-daemon/pull/32) ([mpetrun5](https://github.com/mpetrun5))
- Added integration with load balancer http tunnel [\#39](https://github.com/NodeFactoryIo/vedran-daemon/pull/39) ([mpetrun5](https://github.com/mpetrun5))

### Changed
- Remove sentry [\#35](https://github.com/NodeFactoryIo/vedran-daemon/pull/35) ([mpetrun5](https://github.com/mpetrun5))

## [v0.1.2](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.1.2)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.1.1...v0.1.2)

### Fixed
- Retry config hash retrieval on fail [\#30](https://github.com/NodeFactoryIo/vedran-daemon/pull/30) ([mpetrun5](https://github.com/mpetrun5))

### Added

### Changed

## [v0.1.1](https://github.com/NodeFactoryIo/vedran-daemon/tree/v0.1.1)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/v0.1.0...v0.1.1)

### Fixed
- Fix ReadyTransactionCount json field [\#27](https://github.com/NodeFactoryIo/vedran-daemon/pull/27) ([mpetrun5](https://github.com/mpetrun5))

### Added

### Changed

## [v0.1.0](https://github.com/NodeFactoryIo/vedran-daemon/tree/v.0.1.0)

[Full Changelog](https://github.com/NodeFactoryIo/vedran-daemon/compare/00a487dbbbd8e201b1354080e88302e0db4130f0...v.0.1.0)

### Fixed
- Fix publish condition [\#20](https://github.com/NodeFactoryIo/vedran-daemon/pull/20) ([mpetrun5](https://github.com/mpetrun5))
- Repo cleanup [\#12](https://github.com/NodeFactoryIo/vedran-daemon/pull/12) ([mpetrun5](https://github.com/mpetrun5))

### Added
- Release pipeline [\#18](https://github.com/NodeFactoryIo/vedran-daemon/pull/18) ([mpetrun5](https://github.com/mpetrun5))
- Build matrix [\#17](https://github.com/NodeFactoryIo/vedran-daemon/pull/17) ([mpetrunic](https://github.com/mpetrunic))
- Start command [\#15](https://github.com/NodeFactoryIo/vedran-daemon/pull/15) ([mpetrun5](https://github.com/mpetrun5))
- Periodicall lb calls [\#14](https://github.com/NodeFactoryIo/vedran-daemon/pull/14) ([mpetrun5](https://github.com/mpetrun5))
- Load balancer client with register [\#11](https://github.com/NodeFactoryIo/vedran-daemon/pull/11) ([mpetrun5](https://github.com/mpetrun5))
- Fetching metrics from prometheus server [\#10](https://github.com/NodeFactoryIo/vedran-daemon/pull/10) ([mpetrun5](https://github.com/mpetrun5))
- Initial repo setup [\#7](https://github.com/NodeFactoryIo/vedran-daemon/pull/7) ([mpetrun5](https://github.com/mpetrun5))
- Setup logging [\#21](https://github.com/NodeFactoryIo/vedran-daemon/pull/21) ([mpetrun5](https://github.com/mpetrun5))
- Config hash logic [\#19](https://github.com/NodeFactoryIo/vedran-daemon/pull/19) ([mpetrun5](https://github.com/mpetrun5))

### Changed


cmd/start.go
package cmd

import (
	"fmt"
	"github.com/NodeFactoryIo/vedran-daemon/internal/lb"
	"github.com/NodeFactoryIo/vedran-daemon/internal/node"
	"github.com/NodeFactoryIo/vedran-daemon/internal/run"
	"github.com/NodeFactoryIo/vedran-daemon/internal/telemetry"
	"github.com/NodeFactoryIo/vedran-daemon/internal/tunnel"
	"github.com/NodeFactoryIo/vedran-daemon/pkg/logger"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"net/url"
	"os"
)

var (
	logLevel       string
	logFile        string
	nodeRPCURL     string
	nodeWSURL      string
	nodeMetricsURL string
	id             string
	lbBaseURL      string
	payoutAddress  string
	lbURL          *url.URL
	metricsURL     *url.URL
	rpcURL         *url.URL
	wsURL          *url.URL
)

var startCmd = &cobra.Command{
	Use:   "vedran-daemon",
	Short: "Register vedran-daemon with load balancer and start sending telemetry",
	RunE:  start,
	PreRunE: func(cmd *cobra.Command, args []string) error {
		level, err := log.ParseLevel(logLevel)
		if err != nil {
			log.Fatalf("Invalid log level %s", logLevel)
		}

		err = logger.SetupLogger(level, logFile)
		if err != nil {
			return err
		}

		return nil
	},
	Args: func(cmd *cobra.Command, args []string) error {
		var err error
		lbURL, err = url.Parse(lbBaseURL)
		if err != nil {
			return fmt.Errorf("Failed parsing load balancer url: %v", err)
		}

		metricsURL, err = url.Parse(nodeMetricsURL)
		if err != nil {
			return fmt.Errorf("Failed parsing metrics url: %v", err)
		}

		rpcURL, err = url.Parse(nodeRPCURL)
		if err != nil {
			return fmt.Errorf("Failed parsing rpc url: %v", err)
		}

		wsURL, err = url.Parse(nodeWSURL)
		if err != nil {
			return fmt.Errorf("Failed parsing ws url: %v", err)
		}

		return nil
	},
}

func init() {
	startCmd.Flags().StringVar(&nodeRPCURL, "node-rpc", "http://localhost:9933", "Polkadot node rpc url")
	startCmd.Flags().StringVar(&nodeWSURL, "node-ws", "ws://localhost:9944", "Polkadot node websocket url")
	startCmd.Flags().StringVar(&nodeMetricsURL, "node-metrics", "http://localhost:9615", "Polkadot node metrics url")
	startCmd.Flags().StringVar(&id, "id", "", "Vedran-daemon id string (required)")
	startCmd.Flags().StringVar(&lbBaseURL, "lb", "", "Target load balancer url (required)")
	startCmd.Flags().StringVar(&payoutAddress, "payout-address", "", "Payout address to which reward tokens will be sent (required)")
	startCmd.Flags().StringVar(&logLevel, "log-level", "info", "Level of logging (eg. debug, info, warn, error)")
	startCmd.Flags().StringVar(&logFile, "log-file", "", "Path to logfile. If not set defaults to stdout")

	_ = startCmd.MarkFlagRequired("id")
	_ = startCmd.MarkFlagRequired("lb")
	_ = startCmd.MarkFlagRequired("payout-address")
}

func start(cmd *cobra.Command, _ []string) error {
	DisplayBanner()

	err := checkIfWsAvailable(wsURL)
	if err != nil {
		return err
	}

	lbClient := lb.NewClient(lbURL)
	nodeClient := node.NewClient(rpcURL, metricsURL)
	telemetry := telemetry.NewTelemetry()
	tunnel := &tunnel.Tunnel{
		NodeRPCURL: rpcURL,
		NodeWSURL:  wsURL,
	}

	err = run.Start(tunnel, lbClient, nodeClient, telemetry, id, payoutAddress)
	if err != nil {
		return fmt.Errorf("Failed starting vedran daemon, because of %v", err)
	}

	return nil
}

// Execute runs command
func Execute() {
	if err := startCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
